### ISSUE TITLE
S3 Resource in Sensor different from S3Resource in other places

### ISSUE THREAD

**Issue by Daniel-Vetter-Coverwhale** [dagster-io/dagster#17374]
### Dagster version

❯ dagster --version dagster, version 1.5.4

### What's the issue?

I have a sensor that needs to poll s3. In order to test it I wanted to use a different s3 resource locally vs prod, so I'm passing in a dagster_aws.s3.S3Resource. According to the docs, this should have a get_client method, and that type actually does, but when the sensor is running the actual type passed in appears to be botocore.client.s3. The picture here shows the type assertion for convenience. If you leave out that assert you get an exception saying that s3 object has not method get_client().

<img width="1383" alt="image" src="https://github.com/dagster-io/dagster/assets/120594412/eb2e0860-1f8f-4e76-81cb-558ccf8b792e">

If you leave out that assert you get 

<img width="1348" alt="image" src="https://github.com/dagster-io/dagster/assets/120594412/a6adba19-6379-490a-a85d-365a39571d5b">

When materializing an asset with the same resource input, the type is correct

<img width="1376" alt="image" src="https://github.com/dagster-io/dagster/assets/120594412/54a9d379-ebb9-48c0-a4ae-c2694d0c9832">

### What did you expect to happen?

The sensor would receive an instance of dagster_aws.s3.S3Resource.

### How to reproduce?

Write this to a file, then run that file with dagster dev -f

```
import dagster
import dagster_aws.s3


@dagster.sensor(job=None)
def test_sensor(context: dagster.SensorEvaluationContext, s3: dagster_aws.s3.S3Resource):
    assert isinstance(s3, dagster_aws.s3.S3Resource), f"Type of s3 is {type(s3)}"
    s3_client = s3.get_client()

    # This doesn't actually matter as you'll error out above
    s3_client.list_objects_v2(Bucket="test")


@dagster.asset
def test_asset(context: dagster.OpExecutionContext, s3: dagster_aws.s3.S3Resource):
    context.log.info(f"Type of s3 is {type(s3)}")
    s3_client = s3.get_client()

    s3_client.list_objects_v2(Bucket="test")

    return None


definitions = dagster.Definitions(
    sensors=[test_sensor], assets=[test_asset], resources={"s3": dagster_aws.s3.S3Resource()}
)
```

### Deployment type

Local

### Deployment details

rtx installed python 3.11.1

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by jamiedemaria** [comment]
cc @benpankow I think this is the same pythonic resources bug we've seen before - do you know if there is another issue tracking it?

**Comment by markfickett** [comment]
Seems to be fixed in 1.8.4, I have an asset which just did `s3_resource.list_files_v2` in Dagster 1.4.4 and in 1.8.4 I need the `get_client()` as I would expect. (It was in a repo with a very old pin; not sure about any intervening versions.)
